docs/topics/typecasts.md

```diff
@@ -60,17 +67,47 @@ when (x) {
 }
 ```
 
-Note that smart casts work only when the compiler can guarantee that the variable won't change between the check and the usage.
-More specifically, smart casts can be used under the following conditions:
-
-* `val` local variables - always, with the exception of [local delegated properties](delegated-properties.md).
-* `val` properties - if the property is private or internal or if the check is performed in the same [module](visibility-modifiers.md#modules) where the property is declared. Smart casts cannot be used on open properties or properties that have custom getters.
-* `var` local variables - if the variable is not modified between the check and the usage, is not captured in a lambda that modifies it, and is not a local delegated property.
-* `var` properties - never, because the variable can be modified at any time by other code.
+Note that smart casts work only when the compiler can guarantee that the variable won't change between the check and its usage.
```

Comment:
I think an obligatory requirement like that should be an actual note (warning) or have a separate header